test/e2e/xfn_test.go

```diff
@@ -0,0 +1,318 @@
+		r := resourceValue(t, claim, "spec", "resourceRef")
+
+		xr := rg(r["name"], "default", r["apiVersion"], r["kind"])
+		mrefs := resourceSliceValue(t, xr, "spec", "resourceRefs")
+		for _, mref := range mrefs {
+			err := wait.For(func() (done bool, err error) {
+				mr := rg(mref["name"], "default", mref["apiVersion"], mref["kind"])
+				l, found := mr.GetLabels()[labelName]
+				if !found {
+					return false, nil
+				}
+				if l != "true" {
+					return false, nil
+				}
+				return true, nil
+			}, wait.WithTimeout(5*time.Minute))
+			if err != nil {
+				t.Fatalf("Expected label %v value to be true", labelName)
+			}
+
+		}
+		return ctx
+	}
+}
+
+// CreateTLSCertificateAsSecret for given dns name and store the secret in the given namespace
+func CreateTLSCertificateAsSecret(dnsName string, ns string) features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		caPem, keyPem, err := createCert(dnsName)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-cert",
+				Namespace: ns,
+			},
+			Type: corev1.SecretTypeTLS,
+			StringData: map[string]string{
+				"tls.crt": caPem,
+				"tls.key": keyPem,
+			},
+		}
+		client := config.Client().Resources()
+		if err := client.Create(ctx, secret); err != nil {
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		i := slices.IndexFunc(nodes.Items[0].Status.Addresses, func(a corev1.NodeAddress) bool {
+			return a.Type == corev1.NodeInternalIP
+		})
+		if i == -1 {
+			t.Fatalf("no nodes with private address")
+		}
+		addr := nodes.Items[0].Status.Addresses[i].Address
+		p := gexe.RunProc(fmt.Sprintf("skopeo copy docker-daemon:crossplane-e2e/fn-labelizer:latest docker://%s:32000/fn-labelizer:latest --dest-tls-verify=false", addr)).Wait()
+		out, _ := io.ReadAll(p.Out())
+		t.Logf("skopeo stdout: %s", string(out))
+		if p.ExitCode() != 0 {
+			t.Fatalf("copying image to registry not successful, exit code %v std out %v std err %v", p.ExitCode(), string(out), p.Err())
+		}
+		return ctx
+	}
+}
+
+// InstallDockerRegistry with custom TLS
+func InstallDockerRegistry() features.Func {
```

Comment:
Down the line we could also think to move this to the Setup in TestMain as I can see it useful for many tests, is this already exposing both http and https at the same time or is it forcefully redirecting to https?

Reply:
[Docker registry](https://docs.docker.com/registry/deploying/) can either talk http or https but not both. For the purpose of this tests we needed to configure to listen for https trafic, using a custom TLS certificate.

I would rather keep the main setup as light as possible - if a test does not pull anything from a private registry, then we can avoid its deploying, shortening the testing time and resource usage.

Ofcourse, if we notice that some other tests need registry, we could then define a number crossplane deployments (stable features only, with beta features, with alpha features) and run the tests accordingly.

Reply:
Totally agree